rust/lance/src/index/frag_reuse.rs

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::dataset::optimize::remapping::transpose_row_ids;
+use crate::Dataset;
+use lance_core::Error;
+use lance_index::frag_reuse::{
+    FragReuseIndex, FragReuseIndexDetails, FragReuseVersion, FRAG_REUSE_DETAILS_FILE_NAME,
+    FRAG_REUSE_INDEX_NAME,
+};
+use lance_index::DatasetIndexExt;
+use lance_table::format::pb::fragment_reuse_index_details::{Content, InlineContent};
+use lance_table::format::pb::{ExternalFile, FragmentReuseIndexDetails};
+use lance_table::format::{Fragment, Index};
+use prost::Message;
+use roaring::{RoaringBitmap, RoaringTreemap};
+use snafu::location;
+use std::collections::HashMap;
+use std::io::Cursor;
+use std::sync::Arc;
+use tokio::io::AsyncWriteExt;
+
+/// Load fragment reuse index details from index metadata
+pub async fn load_frag_reuse_index_details(
+    dataset: &Dataset,
+    index: &Index,
+) -> lance_core::Result<Arc<FragReuseIndexDetails>> {
+    let details_any = index.index_details.clone();
+    if details_any.is_none()
+        || !details_any
+            .as_ref()
+            .unwrap()
+            .type_url
+            .ends_with("FragmentReuseIndexDetails")
+    {
+        return Err(Error::Index {
+            message: "Index details is not for the fragment reuse index".into(),
+            location: location!(),
+        });
+    }
+
+    let proto = details_any.unwrap().to_msg::<FragmentReuseIndexDetails>()?;
+    match &proto.content {
+        None => Err(Error::Index {
+            message: "Index details content is not found".into(),
+            location: location!(),
+        }),
+        Some(Content::Inline(content)) => {
+            Ok(Arc::new(FragReuseIndexDetails::try_from(content.clone())?))
+        }
+        Some(Content::External(external_file)) => {
+            let file_path = dataset
+                .base
+                .child(index.uuid.to_string())
+                .child(external_file.path.clone());
+
+            dataset
+                .session
+                .file_metadata_cache
+                .get_or_insert(&file_path, |_path| async {
+                    let range = external_file.offset as usize
+                        ..(external_file.offset as usize + external_file.size as usize);
+                    let data = dataset
+                        .object_store
+                        .open(&file_path)
+                        .await?
+                        .get_range(range)
+                        .await?;
+
+                    let pb_sequence = InlineContent::decode(data)?;
+                    FragReuseIndexDetails::try_from(pb_sequence)
```

Comment:
Are we double caching this data? Won't it also be stored in the index cache?

Reply:
yeah you are right, the content will be in the index cache anyway so this is not necessary. Let me remove